File: consensus/eonix_consensus.py
import hashlib
import time
from typing import List, Optional

from eonix_network.blockchain import Blockchain
from eonix_network.crypto import Crypto
from eonix_network.network import Network
from eonix_network.transaction import Transaction

class EonixConsensus:
    def __init__(self, blockchain: Blockchain, network: Network, crypto: Crypto):
        self.blockchain = blockchain
        self.network = network
        self.crypto = crypto
        self.validators = {}
        self.current_block = None

    def add_validator(self, validator: 'Validator'):
        self.validators[validator.public_key] = validator

    def remove_validator(self, public_key: str):
        del self.validators[public_key]

    def create_new_block(self, transactions: List[Transaction]) -> 'Block':
        block = Block(transactions)
        block_hash = self.crypto.hash(block)
        signature = self.crypto.sign(block_hash, self.validators[list(self.validators.keys())[0]].private_key)
        block.signature = signature
        self.blockchain.add_block(block)
        return block

    def verify_block(self, block: 'Block') -> bool:
        block_hash = self.crypto.hash(block)
        return self.crypto.verify(block_hash, block.signature, self.validators[list(self.validators.keys())[0]].public_key)

    def start_consensus(self):
        while True:
            time.sleep(1)
            current_block = self.current_block
            if current_block is None or time.time() - current_block.timestamp > 10:
                transactions = self.network.get_transactions()
                block = self.create_new_block(transactions)
                self.current_block = block

class Validator:
    def __init__(self, public_key: str, private_key: str):
        self.public_key = public_key
        self.private_key = private_key

class Block:
    def __init__(self, transactions: List[Transaction]):
        self.transactions = transactions
        self.timestamp = time.time()
        self.signature = None

    def __hash__(self):
        return int(hashlib.sha256(str(self.transactions).encode()).hexdigest(), 16)

File: consensus/eonix_consensus.go
package consensus

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/big"
	"sync"
	"time"

	"github.com/eonix-project/eonix-network/blockchain"
	"github.com/eonix-project/eonix-network/crypto"
	"github.com/eonix-project/eonix-network/network"
)

// EonixConsensus represents the Eonix consensus algorithm
type EonixConsensus struct {
	// Blockchain instance
	blockchain *blockchain.Blockchain
	// Network instance
	network *network.Network
	// Crypto instance
	crypto *crypto.Crypto
	// Validator set
	validators []*Validator
	// Current block
	currentBlock *blockchain.Block
	// Lock for concurrent access
	lock sync.RWMutex
}

// NewEonixConsensus creates a new Eonix consensus instance
func NewEonixConsensus(blockchain *blockchain.Blockchain, network *network.Network, crypto *crypto.Crypto) *EonixConsensus {
	return &EonixConsensus{
		blockchain: blockchain,
		network:    network,
		crypto:     crypto,
		validators: make([]*Validator, 0),
	}
}

// AddValidator adds a new validator to the consensus algorithm
func (ec *EonixConsensus) AddValidator(validator *Validator) {
	ec.lock.Lock()
	defer ec.lock.Unlock()
	ec.validators = append(ec.validators, validator)
}

// RemoveValidator removes a validator from the consensus algorithm
func (ec *EonixConsensus) RemoveValidator(validator *Validator) {
	ec.lock.Lock()
	defer ec.lock.Unlock()
	for i, v := range ec.validators {
		if v.Equal(validator) {
			ec.validators = append(ec.validators[:i], ec.validators[i+1:]...)
			return
		}
	}
}

// CreateNewBlock creates a new block and adds it to the blockchain
func (ec *EonixConsensus) CreateNewBlock(transactions []*blockchain.Transaction) (*blockchain.Block, error) {
	// Create a new block with the given transactions
	block := blockchain.NewBlock(transactions)
	// Calculate the block hash
	blockHash := ec.calculateBlockHash(block)
	// Sign the block with the validator's private key
	signature, err := ec.crypto.Sign(blockHash, ec.validators[0].PrivateKey)
	if err != nil {
		return nil, err
	}
	// Add the signature to the block
	block.Signature = signature
	// Add the block to the blockchain
	ec.blockchain.AddBlock(block)
	return block, nil
}

// calculateBlockHash calculates the hash of a block
func (ec *EonixConsensus) calculateBlockHash(block *blockchain.Block) []byte {
	hash := sha256.New()
	hash.Write(block.PrevBlockHash)
	hash.Write(block.TransactionsHash)
	hash.Write(block.Timestamp)
	hash.Write(block.Nonce)
	return hash.Sum(nil)
}

// VerifyBlock verifies a block and its signature
func (ec *EonixConsensus) VerifyBlock(block *blockchain.Block) bool {
	// Calculate the block hash
	blockHash := ec.calculateBlockHash(block)
	// Verify the signature
	return ec.crypto.Verify(blockHash, block.Signature, ec.validators[0].PublicKey)
}

// StartConsensus starts the consensus algorithm
func (ec *EonixConsensus) StartConsensus() {
	go func() {
		for {
			// Get the current block
			ec.lock.RLock()
			currentBlock := ec.currentBlock
			ec.lock.RUnlock()
			// Create a new block if the current block is old
			if currentBlock == nil || time.Since(currentBlock.Timestamp) > 10*time.Second {
				transactions := ec.network.GetTransactions()
				block, err := ec.CreateNewBlock(transactions)
				if err != nil {
					fmt.Println(err)
					continue
				}
				ec.lock.Lock()
				ec.currentBlock = block
				ec.lock.Unlock()
			}
			// Sleep for 1 second
			time.Sleep(1 * time.Second)
		}
	}()
}

// Validator represents a validator in the consensus algorithm
type Validator struct {
	PublicKey  []byte
	PrivateKey []byte
}

// Equal checks if two validators are equal
func (v *Validator) Equal(other *Validator) bool {
	return bytes.Equal(v.PublicKey, other.PublicKey) && bytes.Equal(v.PrivateKey, other.PrivateKey)
}
